### ISSUE TITLE
Update Configuration Contribution Points

### ISSUE THREAD

**Issue by depascalis** [Gruntfuggly/todo-tree#512]
For easier configuration in `settings.json` it would be good if the schema was updated. I'm especially missing properties in `todo-tree.highlights.customHighlight`. It would be good if all properties were included that are mentioned in the documentation under [higlighting](https://github.com/Gruntfuggly/todo-tree#highlighting).

Also, would it be possible to make color properties like `foreground` and `background` to the type of color so that color decorators work? I do not mean `{ "type": "color", ... }` because that doesn't exist, I'm also not referring to `contributes.colors`.

Take for example the default property `screencastMode.mouseIndicatorColor` which has a proper type, and therefore shows a color decorator and color picker.

![image](https://user-images.githubusercontent.com/33980634/118507242-578a6b80-b72e-11eb-9669-e6df047f6579.png)

**Comment by Gruntfuggly** [comment]
I'll have a look at creating JSON schema. It might be a lot of work though 🙁

**Comment by depascalis** [comment]
Awesome, thanks! But is it really that much work? I mean, I thought you just needed to add the missing properties to the manifest, for the first part at least. I'm unsure about the second part though. I was about to contribute myself, but I got confused about how types are handled, the documentation did not have an example on how to implement color types this way.

**Comment by Gruntfuggly** [comment]
The package.json defines most stuff, but the default and custom highlights are more complex objects, so they don't get parsed by the standard vscode package.json schema (as far as I can work out).

I don't know yet if I need to create a schema for the whole package.json (there is a lot of stuff in it!) or whether I can just do part of it.

**Comment by depascalis** [comment]
Ok I think I have figured the first part out. I have no experience with extensions but the following seems to be working. I've set most of it to strings for now. Next is to figure out the color type.

```jsonc
{ 
//...
                  "todo-tree.highlights.customHighlight": {
                    "type": "object",
                    "additionalProperties": {
                        "type": "object",
                        "properties": {
                            "foreground": { "type": "string" },
                            "background": { "type": "string" },
                            "opacity": { "type": "string" },
                            "fontWeight": { "type": "string" },
                            "fontStyle": { "type": "string" },
                            "textDecoration": { "type": "string" },
                            "borderRadius": { "type": "string" },
                            "icon": { "type": "string" },
                            "iconColour": { "type": "string" },
                            "gutterIcon": { "type": "string" },
                            "rulerColour": { "type": "string" },
                            "rulerOpacity": { "type": "string" },
                            "rulerLane": { "type": "string" },
                            "type": { 
                                "type": "string",
                                "enum": [
                                    "tag",
                                    "text",
                                    "tag-and-comment",
                                    "tag-and-subTag",
                                    "text-and-comment",
                                    "line",
                                    "whole-line"
                                ]
                            },
                            "hideFromTree": { "type": "string" },
                            "hideFromStatusBar": { "type": "string" } 
                        }
                    },
                    
                    "default": {
                        "BUG": {
                            "icon": "bug"
                        },
                        "FIXME": {
                            "icon": "flame"
                        }
                    },
                    "markdownDescription": "%todo-tree.configuration.highlights.customHighlight.markdownDescription%"
                },
//...
```

**Comment by depascalis** [comment]
I got colors working too. Only thing left is for some major refactoring of package.json :) I need to focus on my real work now though.

```jsonc
                            "foreground": { 
                                "type": "string",
                                "format": "color-hex",
                                "description": "Color - Supported values are...?"
                            },
```

**Comment by Gruntfuggly** [comment]
Wow - great work! I'll add it for the next release. Where did you find out about the additional properties? and the "format": "color-hex"? I'm not sure it's documented?

**Comment by depascalis** [comment]
Yeah this took a whole day+ due to the lack of documentation and my lack of experience with the JSON Schema and extensions.
I found the `additionalProperties` through IntelliSense where it popped up as a valid property, I then looked it up in the [JSON Schema documentation](https://json-schema.org/understanding-json-schema/reference/object.html#id2). That's why you should have a proper schema ;)
I've been scouring repos for manifests that included "color" and I somehow finally ended up [here](https://github.com/microsoft/vscode/tree/main/extensions/json-language-features/server). I had already been through that documentation, but I never thought of looking under the "server" section. After finding the proper pieces it went pretty fast.

**Comment by Gruntfuggly** [comment]
Thank you - much appreciated!

**Comment by depascalis** [comment]
Be aware of that my types are made in all haste. `opacity` for example is not a number originally, so changing it to a number will breaking change, it should be a number though.

**Comment by Gruntfuggly** [comment]
It seems to work fine - Opacity is treated as a number in the code.

This can only be an enhancement as previously it was down to the user to get it right.
